### PR TITLE
[Feature] Google Cloud storage management #8

### PR DESCRIPTION
**Pull request by davidmeirlevy**

### CODE REVIEW

**Comment by davidmeirlevy** [comment]
@or-yam any updates?

File: server/controllers/assets.js [hunk 1]
```diff
@@ -1,47 +1,52 @@
-const { rename } = require('fs')
-const { loadFiles, removeFile, uploadFile } = require('../services/ftp')
-const { tempFolder } = require('../../config')
+const ftpService = require('../services/ftp')
+const gcsService = require('../services/gcs')
+
+function getService ({ kind }) {
+	if (kind === 'gcs') {
```
Comment (davidmeirlevy) [inline server/controllers/assets.js:5]: there's something wrong with the indentation.

File: server/models/gcs.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    );
+  }
+
+  async list(path) {
+    const options = {
+      prefix: path,
+    };
+    try {
+      const list = await this._client.bucket(this.bucket).getFiles(options);
+      return list;
```
Comment (davidmeirlevy) [inline server/models/gcs.js:31]: why return on a different line? just return the previous line.

File: server/services/gcs.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
+const path = require('path')
+const uniqid = require('uniqid')
+const ASSET_TYPES = require('../../helpers/asset-types.json')
+const { joinUrl } = require('./url')
+const Gcs = require('../models/gcs')
+
+function getAssetType(asset) {
+  return ASSET_TYPES.ASSET;
```
Comment (davidmeirlevy) [inline server/services/gcs.js:8]: a mandatory step is to distinguish between folders, images, and all the rest.

File: server/services/gcs.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    gcs.destroy()
+  }
+
+  return list.map(asset => {
+    const fileIdentifier = path.join(identifier, asset.name)
+
+    return {
+      name: asset.name,
+      identifier: fileIdentifier,
+      updated: asset.date,
```
Comment (davidmeirlevy) [inline server/services/gcs.js:33]: are you sure those properties exist on GCS?

File: server/services/gcs.js (hunk 3 above)
Comment (or-yam) [reply server/services/gcs.js]: tried to solve it like with ftp. i couldn't test it on my environment.

File: server/models/gcs.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      }
+    );
+  }
+
+  async list(path) {
+    const options = {
+      prefix: path,
+    };
+    try {
+      return await this._client.bucket(this.bucket).getFiles(options);
```
Comment (davidmeirlevy) [inline server/models/gcs.js:30]: You stored this.bucket. Why don't you use it?

File: server/models/gcs.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    } catch (error) {
+      return { message: "could not retrieve assets from storage: " + this.name };
+    }
+  }
+
+  async upload(file) {
+    try {
+      await this._client.bucket(this.bucket).upload(file, {
```
Comment (davidmeirlevy) [inline server/models/gcs.js:38]: You stored this.bucket. Why don't you use it?

File: server/models/gcs.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        metadata: {},
+      });
+    } catch (error) {
+      return { message: "could not upload asset to storage: " + this.name };
+    }
+  }
+
+  async remove(file) {
+    try {
+      await this._client.bucket(this.bucket).file(file).delete();
```
Comment (davidmeirlevy) [inline server/models/gcs.js:50]: You stored this.bucket. Why don't you use it?

File: server/services/gcs.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+const ASSET_TYPES = require("../../helpers/asset-types.json");
+const { isImage, joinUrl } = require("./url");
+const Gcs = require("../models/gcs");
+
+function getAssetType(asset) {
+  const type = ASSET_TYPES[asset.type];
+  if (type === ASSET_TYPES.ASSET) {
+    return isImage(asset.name) ? ASSET_TYPES.IMAGE : ASSET_TYPES.ASSET;
+  }
+  return type;
```
Comment (davidmeirlevy) [inline server/services/gcs.js:12]: What about directory type?

File: server/services/gcs.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+const { isImage, joinUrl } = require('./url')
+const Gcs = require('../models/gcs')
+
+function getAssetType(metadata) {
+  return isImage(metadata.name) ? ASSET_TYPES.IMAGE : metadata.kind
+}
+
+async function loadFiles(storage, identifier = '/') {
+  const gcs = new Gcs(storage)
+  const fullPath = path.join(storage.metadata.basePath || '/', identifier)
```
Comment (drorrose) [inline server/services/gcs.js:13]: No need for the OR here, right? The base path parameter has a default of the same value in the record itself. Also, if the identifier is the default, and the base path is the default, then you will see a weird path, so maybe we should not add a default to the identifier?

File: server/services/gcs.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+      type: getAssetType(asset.metadata),
+      publicUrl: joinUrl(storage.metadata.publicUrl, fileIdentifier)
+    }
+  })
+}
+
+async function uploadFile(storage, { identifier, file, extension, prefix }) {
+  const gcs = new Gcs(storage)
+  const filename = `${prefix}-${uniqid()}.${extension}`
+  const fullPath = path.join(storage.metadata.basePath || '/', identifier, filename)
```
Comment (drorrose) [inline server/services/gcs.js:43]: Same comment about the OR

**Comment by davidmeirlevy** [comment]
@drorrose @or-yam I finished viewing the list and add/edit GCS storage credentials.
please finish with the upload and remove, and let's close this issue.

**Pull request merged**